File: neet-code/4.top-k-frequent-elements/python/main.py
from typing import List


class Solution:
    def top_k_frequent_elements(self, nums: List[int], k: int) -> List[int]:
        count = {}
        freq = [[] for i in range(len(nums)+1)]
        print(freq)
        for n in nums:
            count[n] = 1 + count.get(n, 0)
        for n, c in count.items():
            freq[c].append(n)
        res = []
        for i in range(len(freq) - 1, 0, -1):  # go in desencding order
            for n in freq[i]:
                res.append(n)
                if (len(res) == k):
                    return res


soln = Solution()
print(soln.top_k_frequent_elements([1, 1, 1, 2, 2, 4, 4, 4, 3], 2))

File: neet-code/4.top-k-frequent-elements/golang/main.go
package main

import "fmt"

func printTopK(nums []int, k int) int {
	count := make(map[int]int)
	freq := [][]int{}
	for i := 0; i <= len(nums)+1; i++ {
		freq = append(freq, []int{})
	}
	for _, value := range nums {
		count[value]++
	}
	for k, v := range count {
		freq[v] = append(freq[v], k)
	}
	fmt.Println(freq)
	return 0
}

func main() {
	res := printTopK([]int{1, 1, 1, 2, 2, 4, 4, 4, 3}, 2)
	fmt.Println(res)
}
